[BC-17150] As a template contributor, when I add a new methodology template, I should have tests to verify that my templates are against a correct methodology and step

rajkumar-u opened the pull request:
### User story
As a template contributor
When i add a new methodology template
I should have tests to verify that my templates are against a correct methodology and step
So that i can add template without any error

Jira ticket : [BC-17150](https://bugcrowd.atlassian.net/browse/BC-17150) 

### Technical explanation
- Added a test using JSON Schema, that the mapping is valid according to the schema
- Added a test for all templates referred to in the mapping file actually exist
- It will give an error message if template mapping is incorrect or not exist

### To test
1. Run the below commands in your console
```
pip3 install -r lib/requirements.txt
python3 -B lib/validate_bmt.py 
```

2. You can see the valid results
<img width="1088" alt="Screenshot 2021-09-08 at 12 20 33 PM" src="https://user-images.githubusercontent.com/57894121/132460424-6a4087ce-0f7e-4a33-b009-f652208d7c3c.png">

3. To test negative scenarios
In `templates.json` modify the mapping data as invalid
```
{
  "metadata": {
    "default": null
  },
  "content": [
    {
      "methodology": "website_testing",
      "children": [
        {
          "key": "information",
          "attribute": "notes",
          "template": "testing"
        }
      ]
    }
  ]
}
```
4. Again run the script `python3 -B lib/validate_bmt.py `
5. You can see the error message for template missing
<img width="1005" alt="Screenshot 2021-09-08 at 12 22 39 PM" src="https://user-images.githubusercontent.com/57894121/132461015-bcac15a7-8b9a-4149-97b0-6bc554f30e4e.png">
6. kindly check and confirm, when adding or removing new template mappings it should exist.

fedetaglia commented:
hey @rajkumar-u this all make sense, just a couple of questions/suggestions

- have you considered splitting the templates mapping files per methodology? ( no schema changes required )
  ei. not `templates/template.json` but `templates/website_testing.json` etc ..
  
 - is it possible to split the template mapping tests to their own file?

rajkumar-u commented:
> hey @rajkumar-u this all make sense, just a couple of questions/suggestions
> 
> * have you considered splitting the templates mapping files per methodology? ( no schema changes required )
>   ei. not `templates/template.json` but `templates/website_testing.json` etc ..
> * is it possible to split the template mapping tests to their own file?

@fedetaglia as per our discussion, i have added below items
1. added `template.json and template.schema.json` inside the mappings directory(removed templates folder)
2. added `title` in metadata
3. added testing file separately.

arcwhite commented on lib/tests/test_template_mapping.py:
```diff
@@ -0,0 +1,55 @@
+  def check_template_exist(self, methodology, steps):
+      """
+      Check a template mapping path from templates.json file
+      and check methodology template is exist or not in template repository
```
```suggestion
      and check methodology template exists (or not) in template repository
```

arcwhite commented on lib/tests/test_template_mapping.py:
```diff
@@ -0,0 +1,55 @@
+
+  def validate_mapping_data(self, mapping):
+      for methodologyData in mapping:
+          self.check_template_exist(methodologyData['methodology'], methodologyData['children'])
```
```suggestion
          self.check_template_exists(methodologyData['methodology'], methodologyData['children'])
```

arcwhite commented on lib/tests/test_template_mapping.py:
```diff
@@ -0,0 +1,55 @@
+      for methodologyData in mapping:
+          self.check_template_exist(methodologyData['methodology'], methodologyData['children'])
+
+  def check_template_exist(self, methodology, steps):
```
```suggestion
  def check_template_exists(self, methodology, steps):
```
